csharp/src/Ice/OutgoingFrame.cs

```diff
@@ -197,10 +316,63 @@ public CompressionResult CompressPayload()
             _compressionMinSize = compressionMinSize;
         }
 
-        internal void Finish(OutputStream.Position encapsulationEnd)
+        private OutputStream WriteBinaryContextEntryKey(int key)
         {
-            Size = Payload.GetByteCount();
+            if (Protocol == Protocol.Ice1)
+            {
+                throw new NotSupportedException("binary context is not supported with Ice1 protocol");
+            }
+
+            if (IsSealed)
+            {
+                throw new InvalidOperationException("cannot modify a sealed frame");
+            }
+
+            OutputStream.Position encapsulationEnd = _encapsulationEnd ??
+                throw new InvalidOperationException(
+                    "binary context cannot be written before finishing the encapsulation");
+
+            if (_binaryContextKeys.Contains(key))
+            {
+                throw new ArgumentException($"key `{key}' is already in use", nameof(key));
+            }
+
+            var ostr = new OutputStream(Encoding, Payload, _binaryContextEnd ?? encapsulationEnd);
+            if (_binaryContextKeys.Count == 0)
+            {
+                ostr.StartFixedLengthSize(2); // 2-bytes size place holder
+            }
+            ostr.WriteInt(key);
+            _binaryContextKeys.Add(key);
+            return ostr;
+        }
+
+        internal void FinishEncapsulation(OutputStream.Position encapsulationEnd) =>
             _encapsulationEnd = encapsulationEnd;
+
+        internal void FinishPayload(Dictionary<string, string>? context)
```

Comment:
It's not clear to me what this method is about.

In particular, for me:
- the payload does not include the binary context
- the string-string Context does not belong to Outgoing Frame - it's a property of Incoming/Outgoing RequestFrame.

Reply:
This  is doing 3 things:

 * Adjust the list of segments currently named `Payload` to its used size, specially slice the last segment to the last written byte, that need to be done before handling the list to the transport for sending,

* Write the context as the binary context entry 0 if one is provided

* Finish write the size of the binary context

Reply:
Replaced by Finish , that ensures the frame is ready to send. The last segment is adjusted and the binary context size is write.